socorro/cron/jobs/monitoring.py

```diff
@@ -0,0 +1,171 @@
+    required_config.add_option(
+        'node_modules',
+        doc=(
+            'Path to node_modules directory where the webapp\'s npm dependencies have been '
+            'installed.'
+        ),
+    )
+
+    def run(self):
+        vulnerabilities = self.get_python_vulnerabilities() + self.get_javascript_vulnerabilities()
+        if vulnerabilities:
+            try:
+                dsn = self.config.sentry.dsn
+            except KeyError:
+                dsn = None
+
+            if dsn:
+                self.alert_sentry(dsn, vulnerabilities)
+            else:
+                self.alert_log(vulnerabilities)
+
+    def alert_sentry(self, dsn, vulnerabilities):
+        client = raven_client.get_client(dsn)
+        client.context.activate()
+        client.context.merge({
+            'extra': {
+                'data': {vuln.key: vuln.summary for vuln in vulnerabilities},
+            },
+        })
+        client.captureMessage('Dependency security check failed')
+
+    def alert_log(self, vulnerabilities):
+        for vuln in vulnerabilities:
+            self.config.logger.error('%s: %s' % (vuln.key, vuln.summary))
+
+    def get_python_vulnerabilities(self):
+        """Check Python dependencies via Pyup's safety command.
+
+        :returns list(Vulnerability):
+        :raises DependencySecurityCheckFailed:
+        """
+        # Safety checks what's installed in the current virtualenv, so no need
+        # for any paths.
+        process = Popen(['safety', 'check', '--json'], stdin=PIPE, stdout=PIPE, stderr=PIPE)
+        output, error_output = process.communicate()
+        if process.returncode == 0:
+            return []
+        elif process.returncode == 255:
+            try:
+                results = json.loads(output)
+                return [
+                    Vulnerability(
+                        type='python',
+                        dependency=result[0],
+                        installed_version=result[2],
+                        affected_versions=result[1],
+                        description=result[3],
+                    ) for result in results
+                ]
+            except (ValueError, IndexError) as err:
+                raise DependencySecurityCheckFailed(
+                    'Could not parse pyup safety output',
+                    err,
+                    output,
+                )
+
+        raise DependencySecurityCheckFailed(error_output)
+
+    def get_javascript_vulnerabilities(self):
+        """Check JavaScript dependencies via the nsp command.
+
+        :returns list(Vulnerability):
+        :raises DependencySecurityCheckFailed:
+        """
+        process = Popen(
+            [
+                os.path.join(self.config.node_modules, '.bin', 'nsp'),
+                'check',
+                '--reporter=json',
+            ],
+            stdin=PIPE,
+            stdout=PIPE,
+            stderr=PIPE,
+            cwd=os.path.join(REPO_ROOT, 'webapp-django'),
```

Comment:
If `node_modules` is always a subdirectory of `webapp-django`, wouldn't it be better to calculate the cwd with something like:

```
os.path.dirname(self.config.node_modules)
```

?

Then they're tied together config-wise. The -stage and -prod servers use the Socorro RPM and not a git checkout.

Reply:
It is not always a subdirectory of `webapp-django`. In the docker containers it is under `/weapp-frontend-deps/node_modules`.

I can change this to `PROJECT_ROOT` to reflect the lack of a repo in the RPM, though.